File: HW9/task1.cpp
#include <iostream>
#include <vector>

struct Node {
	int y;
	int sz;
	int val;
	int prior;
	Node *l, *r;
};

Node *new_node(int val) {
	Node *result = new Node;
	result->y = std::rand();
	result->sz = 1;
	result->val = val;
	result->l = result->r = nullptr;
	return result;
}

class Treap {
private:
	Node *root;
	int get_sz(Node *t) {
		if (t == nullptr) {
			return 0;
		}
		return t->sz;
	}

	void upd_sz(Node *t) {
		if (t == nullptr) {
			return;
		}
		t->sz = 1 + get_sz(t->l) + get_sz(t->r);
	}

	Node *merge(Node *t1, Node *t2) {
		if (t1 == nullptr) {
			return t2;
		}
		if (t2 == nullptr) {
			return t1;
		}
		if (t1->y > t2->y) {
			t1->r = merge(t1->r, t2);
			upd_sz(t1);
			return t1;
		} else {
			t2->l = merge(t1, t2->l);
			upd_sz(t2);
			return t2;
		}
	}

	void split(Node *t, int x, Node *&t1, Node *&t2) {
		if (t == nullptr) {
			t1 = t2 = nullptr;
			return;
		}
		if (get_sz(t->l) < x) {
			split(t->r, x - get_sz(t->l) - 1, t->r, t2);
			t1 = t;
		} else {
			split(t->l, x, t1, t->l);
			t2 = t;
		}
		upd_sz(t);
	}

	Node *add(Node *t, int pos, int val) {
		Node *t1, *t2;
		split(t, pos, t1, t2);
		Node *new_tree = new_node(val);
		t1 = merge(t1, new_tree);
		return merge(t1, t2);
	}

	Node *remove(Node *t, int pos) {
		Node *t1, *t2, *t3, *t4;
		split(t, pos, t1, t2);
		split(t2, 1, t3, t4);
		t = merge(t1, t4);
		delete t3;
		return t;
	}

	int get_value(Node *t, int pos) {
		int my_idx = get_sz(t->l);
		if (pos < my_idx) {
			return get_value(t->l, pos);
		} else if (pos == my_idx) {
			return t->val;
		} else {
			return get_value(t->r, pos - my_idx - 1);
		}
	}

	Node *to_front(Node *t, int l, int r) {
		Node *t1, *t2, *t3, *t4;
		split(t, r + 1, t1, t2);
		split(t1, l, t3, t4);
		return merge(merge(t4, t3), t2);
	}

	void print_tree(Node *t) {
		if (t == nullptr) {
			return;
		}
		print_tree(t->l);
		std::cout << t->val << " ";
		print_tree(t->r);
	}

	void insert(Node *& t, Node *it) {
		if (!t)
			t = it;
		else if (it->prior > t->prior)
			split (t, it->val, it->l, it->r),  t = it;
		else
			insert (it->val < t->val ? t->l : t->r, it);
	}

	void erase (Node *& t, int key) {
		if (t->val == key) {
			t = merge( t->l, t->r);
		}
		else {
			erase (key < t->val ? t->l : t->r, key);
		}
	}

	std::string next(Node *t, int key) {
		std::string res = "none";
		while (t != nullptr) {
			if (t->val <= key) {
				t = t->r;
			} else if (t->val > key) {
				res = std::to_string(t->val);
				t = t->l;
			}
		}
		return res;
	}

	std::string prev(Node *t, int key) {
		std::string res = "none";
		while (t != nullptr) {
			if (t->val >= key) {
				t = t->l;
			} else if (t->val < key) {
				res = std::to_string(t->val);
				t = t->r;
			}
		}
		return res;
	}

	std::string search(Node *t, int key) {
		if (t == nullptr) {
			return "false";
		}
		if (t->val == key) {
			return "true";
		} else if (key < t->val) {
			return search(t->l, key);
		} else {
			return search(t->r, key);
		}
	}

public:
	Treap() { root = nullptr; }

	void from_vector(const std::vector<int> &v) {
		Node *result = nullptr;
		for (int i : v) {
			result = merge(result, new_node(i));
		}
		this->root = result;
	}

	void add(int pos, int val) {
		this->root = add(this->root, pos, val);
	}

	void remove(int pos) {
		this->root = remove(this->root, pos);
	}

	void print_tree() {
		print_tree(this->root);
	}

	int get_sz() {
		return get_sz(this->root);
	}

	void to_front(int l, int r) {
		this->root = to_front(this->root, l, r);
	}

	void insert(Node *it) {
		if (this->search(this->root, it->val) == "none") {
			insert(this->root, it);
		}
	}

	void erase(int key) {
		erase(this->root, key);
	}

	std::string next(int key) {
		return next(this->root, key);
	}

	std::string prev(int key) {
		return prev(this->root, key);
	}

	std::string search(int key) {
		return search(this->root, key);
	}
};


int main() {
	std::string str;
	int value;


	Treap treap = Treap();

	while (std::cin >> str) {
		std::cin >> value;
		if (str == "insert") {
			treap.insert(new_node(value));
		} else if (str == "exists") {
			std::cout << treap.search(value) << "\n";
		} else if (str == "next") {
			std::cout << treap.next(value) << "\n";
		} else if (str == "prev") {
			std::cout << treap.prev(value) << "\n";
		} else if (str == "delete") {
			treap.erase(value);
		}
		treap.print_tree();
		std::cout << "\n";
	}

	return 0;
}

File: HW9/task1.py
import random
import sys

SEPARATOR = "\n"
UNICODE = "utf-8"
MAX_PRIOR = 1000

class Node(object):
    def __init__(self, key, prior):
        self.key = key
        self.prior = prior
        self.parent = None
        self.l = None
        self.r = None


class Treap(object):
    nodes = set()

    @staticmethod
    def gen():
        return random.randint(1, MAX_PRIOR)

    def __init__(self):
        self.root = None

    def split(self, t, key):
        if t is None:
            return (None, None)
        if key < t.key:
            l, t.l = self.split(t.l, key)
            r = t
            return (l, r)
        else:
            t.r, r = self.split(t.r, key)
            l = t
            return (l, r)

    def merge(self, l, r):
        if l is None:
            return r
        elif r is None:
            return l
        elif l.prior > r.prior:
            l.r = self.merge(l.r, r)
            return l
        else:
            r.l = self.merge(l, r.l)
            return r

    def find(self, t, key):
        if t is None:
            return "false"
        if t.key == key:
            return "true"
        elif key < t.key:
            return self.find(t.l, key)
        else:
            return self.find(t.r, key)

    def next(self, key):
        res = "none"
        t = self.root
        while t:
            if t.key <= key:
                t = t.r
            elif t.key > key:
                res = t.key
                t = t.l
        return str(res)

    def prev(self, key):
        res = "none"
        t = self.root
        while t:
            if t.key < key:
                res = t.key
                t = t.r
            elif t.key >= key:
                t = t.l
        return str(res)

    def insert(self, key):
        if key in self.nodes:
            return
        self.nodes.add(key)
        new_node = Node(key, self.gen())
        if self.root is None:
            self.root = new_node
            return
        l, r = self.split(self.root, key - 1)
        self.root = self.merge(l, new_node)
        self.root = self.merge(self.root, r)

    def remove(self, key):
        if key not in self.nodes:
            return
        self.nodes.remove(key)
        l, r = self.split(self.root, key)
        l1, _ = self.split(l, key - 1)
        self.root = self.merge(l1, r)

    def get_height(self, t):
        if t is None:
            return 0
        ans = 1
        ans = max(ans, self.get_height(t.l) + 1)
        ans = max(ans, self.get_height(t.r) + 1)
        return ans

    def pr(self, t):
        if t is None:
            return
        if t.l:
            self.pr(t.l)
        print(t.key, end=" ")
        if t.r:
            self.pr(t.r)


if __name__ == "__main__":
    treap = Treap()
    data = sys.stdin.buffer.readlines()
    result_exists = []
    n_operations = len(data)
    for i in range(n_operations):
        command, value = data[i].decode(UNICODE).split()
        if command == "insert":
            treap.insert(int(value))
        elif command == "delete":
            treap.remove(int(value))
        elif command == "exists":
            result_exists.append(treap.find(treap.root, int(value)))
        elif command == "next":
            result_exists.append(treap.next(int(value)))
        elif command == "prev":
            result_exists.append(treap.prev(int(value)))
    encoded_array = (SEPARATOR.join(result_exists)).encode(UNICODE)
    sys.stdout.buffer.write(encoded_array)
